### ISSUE TITLE
Universal binary iree-compile do not run on M1

### ISSUE THREAD

**Issue by raikonenfnu** [openxla/iree#9028]
### Problem: Trying to run e2e Resnet model on M1 with preinstalled binaries but ran into device not supported.

```bash
RuntimeError: Error invoking function: <vm>:0: UNKNOWN; device not supported in the compiled configuration;
[ 0] bytecode module.forward:1996 <stdin>:279:10
```

#### Further investigation show that issue stem from iree-compile
```bash
/Users/user/venv/shark/lib/python3.9/site-packages/iree/compiler/_mlir_libs/iree-compile  --iree-mlir-to-vm-bytecode-module -iree-hal-target-backends=dylib /tmp/Resnet50Module.mlir -iree-vm-bytecode-source-listing=foo.mlir -o Resnet50Module-dylib.vmfb

iree-build/iree/tools/iree-benchmark-module --driver=dylib --entry_function=forward --function_input=1x3x224x224xf32 --module_file=Resnet50Module-dylib.vmfb --benchmark_repetitions=1

# Result:
Running ../iree-build/iree/tools/iree-benchmark-module
Run on (8 X 24.2407 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB (x8)
  L1 Instruction 128 KiB (x8)
  L2 Unified 4096 KiB (x2)
Load Average: 1.58, 1.52, 1.51
<vm>:0: UNKNOWN; device not supported in the compiled configuration;
[ 0] bytecode module.forward:1996 [
    /tmp/Resnet50Module.mlir:279:10
      at /tmp/Resnet50Module.mlir:9:3,
    "vm"(foo.mlir:295:4)
  ]
[1]    84694 abort      ../iree-build/iree/tools/iree-benchmark-module --driver=vulkan
```
But when I use iree-compile built from source it can translate and run fine.

Looking at the iree-vm-bytecode-source-listing seems like the **preinstalled binary iree-compile** is generating **x86**:
<img width="487" alt="Screen Shot 2022-04-29 at 9 51 22 AM" src="https://user-images.githubusercontent.com/68087699/165996405-9caffd13-a682-41b4-923d-25d172dd3509.png">

While if we run **iree-compile built from source** it will generate **ARM**:
<img width="510" alt="Screen Shot 2022-04-29 at 9 49 59 AM" src="https://user-images.githubusercontent.com/68087699/165996224-23cd8771-742c-4fd8-9733-7809b2a86259.png">

I was wondering if there is a way to set the iree-compile to generate ARM instead of x86?

**Comment by ScottTodd** [comment]
Interesting that the binary from the python packages and the from source build are differing here.

You can set `-iree-llvm-target-triple` to target any machine (from the LLVM targets that have been compiled in / initialized, see the bottom of [LLVMAOTTarget.cpp](https://github.com/google/iree/blob/aae314e8bb72115aa00ce2556fbd8ed1559d459d/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp)): https://github.com/google/iree/blob/aae314e8bb72115aa00ce2556fbd8ed1559d459d/iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp#L66-L68 (also see the other options in that file)

For example, I've set `-iree-llvm-target-triple=x86_64-linux-gnu` to compile for Linux from Windows.

If you don't set those options explicitly, they'll default to what LLVM can infer: https://github.com/google/iree/blob/aae314e8bb72115aa00ce2556fbd8ed1559d459d/iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp#L28-L30

**Comment by raikonenfnu** [comment]
Thanks @ScottTodd! setting `-iree-llvm-target-triple=arm64-apple-darwin21.4.0` got it working! 

```
Interesting that the binary from the python packages and the from source build are differing here.
```
It is peculiar that source built infer Arm, and prebuilt infer X86 😅 .  Perhaps `llvm::sys::getDefaultTargetTriple()` automatically load the clang it was built on's defaults?

**Comment by powderluv** [comment]
@stellaraccident fyi since she has her M1 now.

**Comment by stellaraccident** [comment]
> It is peculiar that source built infer Arm, and prebuilt infer X86  . Perhaps `llvm::sys::getDefaultTargetTriple()` automatically load the clang it was built on's defaults?

I am guessing that is exactly what it is doing.

**Comment by jpienaar** [comment]
llvm::sys::getDefaultTargetTriple() `returns the default target triple the compiler has been configured to produce code for` and in particular it is set at the point where iree-compile is compiled (so for universal binary I'm assuming that during compilation time it only gets set for x86 and that same config used both sides). We could potentially use `getProcessTriple` instead which should give the triple of the machine being run on [I was trying to verify that but kept running into other issues today].

**Comment by RoboTux** [comment]
AFAIK this has been fixed in [abc3ac999ea3a2cdea9fdb920dbf368e708eeef7](https://github.com/openxla/iree/commit/abc3ac999ea3a2cdea9fdb920dbf368e708eeef7).

**Comment by ScottTodd** [comment]
Looks like it. Please re-open if there are still issues.
